### DISCUSSION TITLE
How to set `Style` to `Expander.Border`

### DISCUSSION THREAD

**Discussion by Coloryr** [AvaloniaUI/Avalonia#13418] [Q&A]
![image](https://github.com/AvaloniaUI/Avalonia/assets/26276037/a14d930f-8765-4f80-ae87-d01f6b1f697b)

**Comment by thevortexcloud** [comment]
Are you trying to change the border of the content in the expander or the expander's border? If it's the former that is not correct. You should not be using `/template/`. That's for modifying the control template.  Instead just target your border inside the expander. If it's the latter that's also not correct. The control has proper template bindings which means you can just set the border properties on it without a `/template/` selector.

https://github.com/AvaloniaUI/Avalonia/blob/fcea3453ff62225c4be56cf2569edffc72e7402e/src/Avalonia.Themes.Fluent/Controls/Expander.xaml#L81C16-L82C69.

**Comment by robloo** [comment]
It's best to use lightweight styling resources for this if you are modifying the header-area.

ExpanderHeaderBackground
ExpanderHeaderBorderBrush

If you are modifying the content-area just use the standard Background and BorderBrush properties. Comments above are correct -- you need to fix the selector.

**Reply by Coloryr** [reply]
How to modify it?
Use style?

**Reply by thevortexcloud** [reply]
Modify what? The expander border?

**Reply by Coloryr** [reply]
> Modify what? The expander border?

yes
The ExpanderHeaderBorderBrush

**Reply by timunie** [reply]
I'd just override the dynamic resource in `App.Resources`. If you use the Fluent Theme it would be: 

```xml
<SolidColorBush x:Key="ExpanderHeaderBorderBrush">Red</SolidColorBrush>
```

Check original styles for the keys to override
https://github.com/AvaloniaUI/Avalonia/blob/fcea3453ff62225c4be56cf2569edffc72e7402e/src/Avalonia.Themes.Fluent/Controls/Expander.xaml#L69-L69

**Reply by ISDHN** [reply]
> I'd just override the dynamic resource in `App.Resources`. If you use the Fluent Theme it would be:
> 
> ```
> <SolidColorBush x:Key="ExpanderHeaderBorderBrush">Red</SolidColorBrush>
> ```
> 
> Check original styles for the keys to override
> 
> https://github.com/AvaloniaUI/Avalonia/blob/fcea3453ff62225c4be56cf2569edffc72e7402e/src/Avalonia.Themes.Fluent/Controls/Expander.xaml#L69-L69

May I asked why neither the Style nor the ControlTheme works for this?

**Reply by robloo** [reply]
> May I asked why neither the Style nor the ControlTheme works for this? 

What do you mean?
 1. Style will work to change this if you target the nested border using a selector
 2. The Expander control itself only has one border property like other controls. It's more intuitive for that to target the content area instead so we can't use it here.
 3. I gave you the lightweight styling resource to modify this without retemplating which was agreed with Microsoft years ago. We follow upstream WinUI here.

What exactly doesn't work?

**Reply by ISDHN** [reply]
> What do you mean?
> 
> 1. Style will work to change this if you target the nested border using a selector
> 2. The Expander control itself only has one border property like other controls. It's more intuitive for that to target the content area instead so we can't use it here.
> 3. I gave you the lightweight styling resource to modify this without retemplating which was agreed with Microsoft years ago. We follow upstream WinUI here.
> 
> What exactly doesn't work?

Both

```xml
<Style Selector="Expander /template/ Border#ToggleButtonBackground">
	<Setter Property="Background" Value="#2F2F2F"/>
</Style>
```
in Application.Styles
and
```xml
<ControlTheme x:Key="{x:Type Expander}" BasedOn="{StaticResource {x:Type Expander}}" TargetType="Expander">
	<Style Selector="^ /template/ Border#ToggleButtonBackground">
		<Setter Property="Background" Value="#2F2F2F"/>
	</Style>
</ControlTheme>
```
in Application.Resource doesn't work.

**Reply by robloo** [reply]
It's nested, you have to target the ToggleButton control template inside the Expander control template itself.

That said, the control theme changes dynamically so that may be an undefined edge case. I would still expect it to work though if you don't change the expand direction after initialization.

https://github.com/AvaloniaUI/Avalonia/blob/f7feaba5a093698e8935ee1caee9831003da2f78/src/Avalonia.Themes.Fluent/Controls/Expander.xaml#L312

**Reply by ISDHN** [reply]
Ok, I see. Thank you.
